### PR TITLE
Initial iostream class to read object media.

### PR DESCRIPTION
**Pull request by coryan**
This is the first version, it just reads the first part of the
object.  Good enough to write tests and validate the API.
Part of the work for #554.

### CODE REVIEW

File: google/cloud/storage/object_stream.cc [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+      return RepositionInputSequence();
+    }
+    // TODO(#714) - use policies to decide if the operation is idempotent.
+    // TODO(#581) - use policies to determine what error codes are permanent.
+    if (not IsRetryableStatusCode(last_status.status_code())) {
+      std::ostringstream os;
+      os << "Permanent error in " << __func__ << ": " << last_status;
+      google::cloud::internal::RaiseRuntimeError(os.str());
+    }
+    std::this_thread::sleep_for(std::chrono::milliseconds(100));
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/object_stream.cc]: Isnt't that covered by the TODO in line 45?

File: google/cloud/storage/internal/default_client.h [hunk 2]
```diff
@@ -79,6 +79,26 @@ class DefaultClient : public Client {
...
 
+  std::pair<Status, std::string> ReadObjectRangeMedia(
+      internal::ReadObjectRangeRequest const& request) override {
+    // Assume the bucket name is validated by the caller.
+    HttpRequest http_request(storage_endpoint_ + "/b/" + request.bucket_name() +
+                             "/o/" + request.object_name());
+    http_request.AddQueryParameter("alt", "media");
+    request.AddParametersToHttpRequest(http_request);
+    http_request.AddHeader(options_.credentials()->AuthorizationHeader());
+    http_request.AddHeader("Range: bytes=" + std::to_string(request.begin()) +
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/internal/default_client.h]: Done.

File: google/cloud/storage/object_stream.cc [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+              "storage::ObjectReadStream must be move constructible.");
+
+std::streamsize ObjectReadStreamBuf::showmanyc() {
+  // TODO(#554) - use the returned headers to have a better estimate.
+  return 0;
+}
+
+ObjectReadStreamBuf::int_type ObjectReadStreamBuf::underflow() {
+  // TODO(#554) - for now, just read the first 64KiB
+  std::int64_t const max_read_bytes = 64 * 1024;
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/object_stream.cc]: We could define something like this:

```C++
namespace byte_literals {
constexpr std::int64_t operator"" _MiB (unsigned long long n) {
  return n * 1024 * 1024;
}
}

...

using namespace google::cloud::byte_literals;
auto size = 64_MiB;
```

For time, there are things `std::chrono::seconds(std::chrono::hours(1)).count()` gives you the number of seconds in an hour, or (in C++14) `std::chrono::seconds(1h).count()`.

**Comment by coryan** [comment]
@houglum PTAL.

File: google/cloud/storage/object_stream.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (houglum) [reply google/cloud/storage/object_stream.cc]: While that does do the math, it's still not 100% clear (to someone reading it for the first time) what the actual value ends up being... especially with the `auto` in front of it.

    // "So... is this the int 64? or 67,108,864? or some special _MiB type?
    // I'll have to go look up _MiB to find out"
    auto size = 64_MiB;

Anyway, not *super* important. Just something to keep in mind as we start doing more of the unit conversions.

Also, thanks for pointing out the `std::chrono` stuff!

File: google/cloud/storage/object_stream.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (houglum) [reply google/cloud/storage/object_stream.cc]: I'm either blind, or GitHub collapsed that line when I was reviewing previously :)

**Pull request merged**